### PR TITLE
Test-Driven Development

### PR DESCRIPTION
**Pull request by omermadmon**
This PR will focus on fixing and optimizing the portfolio project, as well as working on a separated calendar project as part of the last week of the training, focusing on TDD.
 
* Calendar project (TDD):
  * Write query logic to pass existing tests (for mandatory attendees only).
  * Add some more tests for optional attendees feature (including optional attendees when possible).
  * Expand query logic by passing those tests.
* Portfolio project:
  * Create a single class-wide DatastoreService field instead of creating a new instance of DatastoreService just-in-time (in each relevant servlet class).
  * Add comment and achievement constructors to get a datastore entity rather than each field separately.

### CODE REVIEW

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 1]
```diff
@@ -16,18 +16,35 @@
...
+import org.junit.BeforeClass;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
 
 @RunWith(JUnit4.class)
 public final class GreeterTest {
+  private static Greeter greeter;
+  private static String greeting;
+
+  @BeforeClass
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java:28]: In tests, recommended practice *encourages* repetition - it's preferred that each testcase include its own setup ("arrange" section), invoke the method under test ("act" section), and perform any required assertions.
This way, each testcase can be read on its own, without needing to refer elsewhere to understand why it works.

@Before (or @BeforeClass) should be used only for setting up global "scaffolding", not relevant to the code under test itself: e.g., you may use these to start a database service, which the class under test then uses.

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 2]
```diff
@@ -16,18 +16,35 @@
...
+  public static void initGreeter() {
+    greeter = new Greeter();
+    greeting = greeter.greet("Ada");
+  }
 
   @Test
   public void testGreeting() {
-    Greeter greeter = new Greeter();
+    Assert.assertEquals("Hello Ada", greeting);
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java:36]: Google unit-test practices recommend to use Truth (https://truth.dev) rather than JUnit assertions.

Truth assertions are more "fluent", have fewer corner-cases, and print out better error messages.

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 3]
```diff
@@ -16,18 +16,35 @@
...
+  }
 
-    String greeting = greeter.greet("Ada");
+  @Test
+  public void testGreetingTrimsWhitespace() {
+  // Whitespace should be trimmed
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java:41]: The testcase name already says this; no need to have a comment as well.
(Google style recommendation is to avoid comments unless there really is no way to express the same in code: experience shows that engineers will evolve the code and leave obsolete comments behind.)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,121 @@
...
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Create mandatory, optional and all attendees lists.
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    ArrayList<String> allAttendees = new ArrayList<String>();
+    allAttendees.addAll(mandatoryAttendees);
+    allAttendees.addAll(optionalAttendees);
+
+    // if there duration is longer than a day, return an empty list
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:36]: "there duration" --> "the dueation"?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,121 @@
...
+    if (request.getDuration() > NUMBER_OF_MINUTES_IN_DAY) {
+        return Arrays.asList();
+    }
+    
+    // if there are no events or no (non-optional) attendees, return {@code TimeRange.WHOLE_DAY} as a list
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:41]: Is this special-case logic really needed? Won't the logic below return the same result anyway?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,121 @@
...
+    if (allAttendees.isEmpty() || events.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    /** create two lists: one for time ranges of relevant events w.r.t all attendees,
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:46]: Avoid using JavaDoc-style comments (`/** ... */`) inside code. Prefer to use multiple `//`-style comments.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,121 @@
...
+        and one for time ranges of relevant events w.r.t only mandatory attendees. */
+    ArrayList<ArrayList<TimeRange>> relevantTimeRangesLists =
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:48]: Instead of using this nested structure, consider just having a simple method that returns a single list - and just invoke it twice (once with all attendees, once with only mandatory ones).

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,121 @@
...
+        and one for time ranges of relevant events w.r.t only mandatory attendees. */
+    ArrayList<ArrayList<TimeRange>> relevantTimeRangesLists = 
+                FindMeetingQuery.filterIrrelevantTimesRanges(events, allAttendees, mandatoryAttendees);
+    
+    ArrayList<TimeRange> relevantTimeRangesAll = relevantTimeRangesLists.get(0);
+    ArrayList<TimeRange> relevantTimeRangesMandatory = relevantTimeRangesLists.get(1);
+
+    // try to consider all attendees. If failed, consider only mandatory attendees.
+    Collections.sort(relevantTimeRangesAll, TimeRange.ORDER_BY_START);
+    Collection<TimeRange> avilableTimeRangesAll =
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:56]: Typo ("avilable" -> "available")

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,121 @@
...
+    
+    // if there are no events or no (non-optional) attendees, return {@code TimeRange.WHOLE_DAY} as a list
+    if (allAttendees.isEmpty() || events.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    /** create two lists: one for time ranges of relevant events w.r.t all attendees,
+        and one for time ranges of relevant events w.r.t only mandatory attendees. */
+    ArrayList<ArrayList<TimeRange>> relevantTimeRangesLists = 
+                FindMeetingQuery.filterIrrelevantTimesRanges(events, allAttendees, mandatoryAttendees);
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: No need to include the class name here (the method is in the same class as the caller).

Comment (nolaviz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: (same applies further down as well)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,121 @@
...
+
+      resultLists.add(relevantTimeRangesAll);
+      resultLists.add(relevantTimeRangesMandatory);
+
+      return resultLists;
+  }
+
+  /** Return true iff at least one of the event's attendees is in the requested attendees. */
+  private static boolean isRelevant(Event event, Collection<String> attendees) {
+      for (String requestedAttendee : attendees) {
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:88]: Collections.disjoint() is your friend. No need to reinvent it here :-)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,121 @@
...
+      return a list of time ranges where all meeting attendees are avilable. */
+  private static Collection<TimeRange> findAvilableTimesRanges(ArrayList<TimeRange> occupiedSlots, long meetingDuration) {
+      
+      if (occupiedSlots.isEmpty()) {
+          return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+      
+      ArrayList<TimeRange> result = new ArrayList<TimeRange>();
+
+      // init start and end of the next avilable slot (to add to result).
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:107]: Typo: avilable -> available

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -14,10 +14,121 @@
...
+      int start = TimeRange.START_OF_DAY;
+      int end;
+
+      for (TimeRange occupiedSlot : occupiedSlots) {
+
+          end = occupiedSlot.start();
+          if (meetingDuration <= end - start) {
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:114]: Can you just use TimeRange.overlaps()?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 13]
```diff
@@ -270,5 +272,163 @@ public void notEnoughRoom() {
...
+    // Options : |--1--|     |--2--|     |--3--|
+
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 3", TimeRange.fromStartDuration(TimeRange.START_OF_DAY, TimeRange.WHOLE_DAY.duration()),
+            Arrays.asList(PERSON_C)));
+
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:297]: No need for more than one blank line.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 14]
```diff
@@ -270,5 +272,163 @@ public void notEnoughRoom() {
...
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:304]: Prefer to have a blank line here: line 303 is the "act" section, while 304 more properly belongs with the assertion phase.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 15]
```diff
@@ -270,5 +272,163 @@ public void notEnoughRoom() {
...
+        Arrays.asList(TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void onlyOptionalAttendeesWithGap() {
+    // Only two optional attendeess with a gap. 
+    // The gap should be identified and retuned.
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:377]: Typo: returned.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 16]
```diff
@@ -270,5 +272,163 @@ public void notEnoughRoom() {
...
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected = Arrays.asList(TimeRange.WHOLE_DAY);
+
+    Assert.assertEquals(expected, actual);
+  }
 }
 
+
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:430]: Remove unnecessary blank lines.

**Review by nolaviz — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/data/Achievement.java [hunk 17]
```diff
@@ -31,4 +33,12 @@ public Achievement(long id, String title, String club,
     this.geo = geo;
     this.text = text;
   }
+
+  public Achievement(Entity entity) {
```
[the remark this answers is not in the document]
Comment (omermadmon) [reply portfolio/src/main/java/com/google/sps/data/Achievement.java:37]: Done.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 18]
```diff
@@ -31,4 +33,12 @@ public Comment(long id, String firstName, String lastName,
     this.timestamp = timestamp;
     this.text = text;
   }
+
+  public Comment(Entity entity) {
```
[the remark this answers is not in the document]
Comment (omermadmon) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:37]: Done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (omermadmon) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:41]: Result should be the same, but I find it more efficient this way (I'd rather recognize it as an edge case than perform the whole query logic just to find out what was clear from the beginning).

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (omermadmon) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:48]: I find it more efficient this way, because this way I am iterating through events just once.

Asymptotically speaking, the runtime complexity analysis would be the same. However, if in the future we will have not only two level of attendence (mandatory and optional), but N levels, than reducing these iteration sounds like a good idea.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (omermadmon) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:88]: howMuchILoveJavaNow++;

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 12 above)
Comment (omermadmon) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:114]: I guess this should work:

```java
if (TimeRange.fromStartEnd(start, end, false).duration() >= meetingDuration)
```

But it is much less readable in my opinion. Did you think of a more elegant way?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (nolaviz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:41]: I'm not sure I agree about the efficiency concern: I usually recommend to only optimize when (or if) you actually see a performance issue. The special-case logic has additional costs (e.g., if I now change the decision logic, the special-case may become incorrect - not likely in this particular case, but not impossible.)

So - this is OK, but IMO not ideal.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (nolaviz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:48]: Efficiency is a distant second concern to code clarity. I really prefer that the code be simpler - if an optimization complicates the code (which you do here), we might avoid it even at the cost of paying more for resources.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 19]
```diff
@@ -14,10 +14,114 @@
...
+    } else {
+        Collections.sort(relevantTimeRangesMandatory, TimeRange.ORDER_BY_START);
+        return findAvailableTimeRanges(relevantTimeRangesMandatory, request.getDuration());
+    }
+  }
+
+  /** Get all events, a list of all attendees and a list of mandatory attendees only.
+      return a list of two lists: one for time ranges of relevant events w.r.t all attendees,
+      and one for time ranges of relevant events w.r.t only mandatory attendees. */
+  private static ArrayList<ArrayList<TimeRange>> filterIrrelevantTimesRanges(Collection<Event> events, Collection<String> allAttendees, Collection<String> mandatoryAttendees) {
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:68]: Note - Google code style recommends a column limit of 100: https://google.github.io/styleguide/javaguide.html#s4.4-column-limit

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 12 above)
Comment (nolaviz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:107]: I actually see the alternative as more readable, mainly because it's obvious that both sides are durations.
But - your call.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (omermadmon) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:41]: Thanks for this comment, will take these considerations into account next time. 
Fixed.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 12 above)
Comment (omermadmon) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:107]: Done.

File: portfolio/src/main/java/com/google/sps/data/Achievement.java [hunk 20]
```diff
@@ -23,6 +25,15 @@
   private final String geo;
   private final String text;
 
+  /**
+  * A constructor who gets the specific fields values of the achievement.
+  * Is being used whenever an achievement object has to be created manually.
```
[the remark this answers is not in the document]
Comment (omermadmon) [reply portfolio/src/main/java/com/google/sps/data/Achievement.java:30]: Fixed in both Comment and Achievement classes.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 21]
```diff
@@ -14,10 +14,94 @@
...
+    allAttendees.addAll(optionalAttendees);
+
+    // if the duration is longer than a day, return an empty list
+    if (request.getDuration() > NUMBER_OF_MINUTES_IN_DAY) {
+        return Arrays.asList();
+    }
+
+    // try to consider all attendees. If failed, consider only mandatory attendees.
+    ArrayList<TimeRange> relevantTimeRangesAll = filterIrrelevantTimesRanges(events, allAttendees);
+    Collections.sort(relevantTimeRangesAll, TimeRange.ORDER_BY_START);
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:43]: Consider folding the sorting into filterIrrelevantTimesRanges (or, alternatively, into findAvailableTimeRanges).
Performing the sort here assumes use of mutable containers, which is against best practices (go/ej3e-17). Also - it's not clear if you're sorting because findAvailableTimeRanges requires it (it has no documentation to say so) or just because you want the output to look nice.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 22]
```diff
@@ -14,10 +14,94 @@
...
+    }
+  }
+
+  /** Get all events and a list of attendees.
+      return a list of time ranges of relevant events w.r.t attendees. */
+  private static ArrayList<TimeRange> filterIrrelevantTimesRanges(Collection<Event> events, Collection<String> attendees) {      
+      ArrayList<TimeRange> relevantTimeRanges = new ArrayList<TimeRange>();
+
+      for (Event event : events) {
+          if (FindMeetingQuery.isRelevant(event, attendees)) relevantTimeRanges.add(event.getWhen());
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:60]: No need to call out the class name for a method in the same class.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 23]
```diff
@@ -14,10 +14,94 @@
...
+        return findAvailableTimeRanges(relevantTimeRangesMandatory, request.getDuration());
+    }
+  }
+
+  /** Get all events and a list of attendees.
+      return a list of time ranges of relevant events w.r.t attendees. */
+  private static ArrayList<TimeRange> filterIrrelevantTimesRanges(Collection<Event> events, Collection<String> attendees) {      
+      ArrayList<TimeRange> relevantTimeRanges = new ArrayList<TimeRange>();
+
+      for (Event event : events) {
```
Comment (nolaviz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:59]: You can simplify this loop using Java8 streams:

  List<TimeRange> relevantTimeRanges = events.stream()
      .filter(event -> isRelevant(event, attendees))
      .collect(Collectors.toList());

**Review by nolaviz — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 23 above)
Comment (omermadmon) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: That's super cool, I'll use it more!
